lib/vrl/DESIGN.md

```diff
@@ -0,0 +1,219 @@
+
+- If useability or readability is hurt by composition, favor single-purpose
+  functions.
+
+#### Naming
+
+- Function names are lower-cased (e.g. `round`).
+
+- Multi-name functions use underscore (`_`) as a separator (e.g.
+  `encode_key_value`).
+
+- Favor explicit verbose names over terse ones (e.g.
+  `parse_aws_cloudwatch_log_subscription_message` over `parse_aws_cwl_msg`).
+
+- Use `parse_*` for string decoding functions (e.g. `parse_json` and
+  `parse_grok`).
+
+- Use `encode_*` for string encoding functions (e.g. `encode_base64`).
+
+- Use `to_*` to convert from one type to another (e.g. `to_bool`).
+
+- Use `format_*` for string formatting functions (e.g. `format_timestamp` and
+  `format_number`).
+
+#### Return Types
+
+- Return boolean from `is_*` functions (e.g. `is_string`).
+
+- Return an error when `parse_*` functions fail to decode the string.
+
+#### Mutability
+
+As a general rule, functions never mutate values.
+
+Favor this:
+
+```coffee
+# explicitly assign the parsed JSON to `.message`
+.message = parse_json(.message)
+```
+
+Over this:
+
+```coffee
+# mutate `.message` in place
+parse_json(.message)
+```
+
+There are exceptions to this rule (such as the `del` function), but they are
+limited, and additional exceptions should be well reasoned.
+
+#### Fallibility
+
+- A function must be marked as fallible if it can fail in any way.
+
+- Design a function with the goal of making it infallible.
+
+- But don't hide fallibility for the sake of persuing the previous rule.
+
+- A function implementation can assume it receives the argument type it has
+  defined.
+
+- `parse_*` functions should almost always error when used incorrectly.
+
+- `to_*` functions must never fail.
+
+#### Signatures
+
+- Functions can have zero or more parameters (e.g. `now()` and `assert(true)`).
+
+- Argument naming follows the same conventions as function naming.
+
+- For one or more parameters, the first parameter must be the "target" of the
+  function (e.g. `parse_regex(target: <string>, pattern: <regex>)`).
+
+- The first parameter must therefor almost always be named `target`.
+
+- TODO
+
+### Errors
+
+TODO
+
+## Patterns
+
+The following is a list of patterns we've experienced while writing and using
+VRL. This section of the document is intended to be updated frequently with new
+insights.
+
+These insights are meant to guide future design decisions, and shape our
+thinking of the language as it matures and we learn more about its strengths and
+weaknesses from our users.
+
+### Error Chaining
+
+Observability data can be structured in unexpected ways. A common pattern is to
+try to decode the data in one way, only to try a different decoder if the first
+one failed.
+
+This pattern uses [function calls][] and [error coalescing][] to achieve its
+goal:
+
+```coffee
+data = parse_json(.message) ??
+       parse_nginx_log(.message) ??
+       parse_apache_log(.message) ??
+       { "error": "invalid data format" }
+```
+
+TODO
```

Comment:
If you've seen any common patterns in the wild that are worth mentioning here, please let me know!

Reply:
👍 I like this pattern.

I am wondering if we'll want to introduce "pipelining" at some point like:

```
data = decode_base64(.message) |> parse_json
```

That is really just syntax sugar though.

Reply:
Pipe operator support is tracked in https://github.com/timberio/vector/issues/5431, but I don't think there's anything to note here, since it isn't actually part of the language yet.